src/kv/RegionCache.cc

```diff
@@ -238,14 +241,27 @@ Store RegionCache::getStore(Backoffer & bo, uint64_t id)
     return reloadStore(store);
 }
 
-std::pair<std::vector<uint64_t>, std::vector<uint64_t>> RegionCache::getAllValidTiFlashStores(Backoffer & bo, const RegionVerID & region_id, const Store & current_store, const LabelFilter & label_filter)
+std::pair<std::vector<uint64_t>, std::vector<uint64_t>> RegionCache::getAllValidTiFlashStores(Backoffer & bo, const RegionVerID & region_id, const Store & current_store, const LabelFilter & label_filter, const std::unordered_set<uint64_t> * store_id_blacklist)
 {
+    auto remove_blacklist = [](const std::unordered_set<uint64_t> * store_id_blacklist, std::vector<uint64_t> & stores, const RegionVerID & region_id, Logger * log){
+        if (store_id_blacklist != nullptr) {
+            auto origin_size = stores.size();
+            stores.erase(std::remove_if(stores.begin(), stores.end(), [&](int x) {
+                return store_id_blacklist->find(x) != store_id_blacklist->end();
+            }), stores.end());
+            if (log != nullptr && origin_size != stores.size()) {
+                auto s = "blacklist peer removed, region=" + region_id.toString() + ", origin_store_size=" + std::to_string(origin_size) + ", current_store_size=" + std::to_string(stores.size());
+                log->information(s);
+            }
+        }
+    };
     std::vector<uint64_t> all_stores;
     std::vector<uint64_t> non_pending_stores;
     RegionPtr cached_region = getRegionByIDFromCache(region_id);
     if (cached_region == nullptr)
     {
         all_stores.emplace_back(current_store.id);
+        remove_blacklist(store_id_blacklist, all_stores, region_id, log);
```

Comment:
After you added this logic, the function `RegionCache::getAllValidTiFlashStores` now may return `all_stores` with an empty set? Can the caller handle this situation?

Reply:
Yes, it is tested in https://github.com/tidbcloud/tiflash-cse/pull/308 where none of the two wn nodes can be selected.